provider/terraform/custom_code.rb

```diff
@@ -0,0 +1,116 @@
+# Copyright 2017 Google Inc.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'api/object'
+require 'provider/abstract_core'
+require 'provider/property_override'
+
+module Provider
+  class Terraform < Provider::AbstractCore
+    # Inserts custom code into terraform resources.
+    class CustomCode < Api::Object
+      # Collection of fields allowed in the CustomCode section for
+      # Terraform.
+
+      # ======================
+      # schema.Resource stuff
+      # ======================
+      # Extra Schema Entries go below all other schema entries in the
+      # resource's Resource.Schema map.  They should be formatted as
+      # entries in the map, e.g. `"foo": &schema.Schema{ ... },`.
+      attr_reader :extra_schema_entry
+      # Resource definition code is inserted below everything else
+      # in the resource's Resource {...} definition.  This may be useful
+      # for things like a MigrateState / SchemaVersion pair.
+      # This is likely to be used rarely and may be removed if all its
+      # use cases are covered in other ways.
+      attr_reader :resource_definition
+      # ====================
+      # Encoders & Decoders
+      # ====================
+      # The encoders are functions which take the `obj` map after it
+      # has been assembled in either "Create" or "Update" and mutate it
+      # before it is sent to the server.  There are lots of reasons you
+      # might want to use these - any differences between local schema
+      # and remote schema will be placed here.
+      # Because the call signature of this function cannot be changed,
+      # the template will place the function header and closing } for
+      # you, and your custom code template should *not* include them.
+      attr_reader :encoder
```

Comment:
This applies to all fields in this file and also the ones in PropertyOverride.

Consider adding the suffix `_template` when the value should be a path to a template and `_func` or `_fn` when it expects a function name.

Reply:
They're all templates - I'll add documentation which says that.